src/components/protocol_handler/src/protocol_packet.cc

```diff
@@ -319,34 +339,45 @@ RESULT_CODE ProtocolPacket::ProtocolHeaderValidator::validate(
   // For Control frames Data Size value shall be less than MTU header
   // For Single and Consecutive Data Size value shall be greater than 0x00
   // and shall be less than payload size
-  if (header.dataSize > payload_size) {
-    LOG4CXX_WARN(logger_,
-                 "Packet data size is "
-                     << header.dataSize
-                     << " and bigger than allowed payload size " << payload_size
-                     << " bytes");
-    return RESULT_FAIL;
-  }
+  // For First Frame Data Size shall be equal 0x08 (payload of this packet will
+  // be 8 bytes).
+
   switch (header.frameType) {
     case FRAME_TYPE_SINGLE:
     case FRAME_TYPE_CONSECUTIVE:
-      if (header.dataSize <= 0u) {
+    case FRAME_TYPE_CONTROL:
+      if (header.dataSize <= 0u || header.dataSize > payload_size) {
+        LOG4CXX_WARN(
+            logger_,
+            "Packet data size of "
+                << StringifyFrameType(header.frameType)
+                << " frame must be in range (0, payload_size=" << payload_size
+                << "], but actual value is " << header.dataSize);
+        return RESULT_FAIL;
+      }
+      break;
+
+    case FRAME_TYPE_FIRST:
+      if (FIRST_FRAME_DATA_SIZE != header.dataSize) {
         LOG4CXX_WARN(logger_,
-                     "Data size of Single and Consecutive frame shall be not "
-                     "equal 0 byte ");
+                     "Packet data size of FRAME_TYPE_FIRST frame must equal "
+                         << int(FIRST_FRAME_DATA_SIZE)
```

Comment:
@KVGrygoriev prefer C++ type cast to C style

Reply:
@AKalinich-Luxoft see e9e7728e3c0b11328d55b083a8bfcbd1542c1f89